sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnSignature.java

```diff
@@ -0,0 +1,275 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import static org.apache.beam.sdk.transforms.DoFn.FinishBundle;
+import static org.apache.beam.sdk.transforms.DoFn.ProcessElement;
+import static org.apache.beam.sdk.transforms.DoFn.StartBundle;
+import static org.apache.beam.sdk.transforms.DoFnReflector.format;
+import static org.apache.beam.sdk.transforms.DoFnReflector.formatType;
+
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+
+import com.google.common.reflect.TypeParameter;
+import com.google.common.reflect.TypeToken;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Describes the signature of a {@link DoFn}, in particular, which features it uses, which extra
+ * context it requires, types of the input and output elements, etc.
+ *
+ * <p>See <a href="https://s.apache.org/a-new-dofn">A new DoFn</a>.
+ */
+class DoFnSignature {
+  public final TypeToken<?> inputT;
+  public final TypeToken<?> outputT;
+  public final ProcessElementMethod processElement;
+  public final BundleMethod startBundle;
+  public final BundleMethod finishBundle;
+
+  private DoFnSignature(
+      TypeToken<?> inputT,
+      TypeToken<?> outputT,
+      ProcessElementMethod processElement,
+      BundleMethod startBundle,
+      BundleMethod finishBundle) {
+    this.inputT = inputT;
+    this.outputT = outputT;
+    this.processElement = processElement;
+    this.startBundle = startBundle;
+    this.finishBundle = finishBundle;
+  }
+
+  /** Describes a {@link DoFn.ProcessElement} method. */
+  public static class ProcessElementMethod {
+    /** The relevant method in the user's class. */
+    public final Method targetMethod;
+    /** 0-based index of the {@link BoundedWindow} parameter, or -1 if it's not used. */
+    public final int boundedWindowParamIndex;
+    /** 0-based index of the {@link DoFn.InputProvider} parameter, or -1 if it's not used. */
+    public final int inputProviderParamIndex;
+    /** 0-based index of the {@link DoFn.OutputReceiver} parameter, or -1 if it's not used. */
+    public final int outputReceiverParamIndex;
+
+    private ProcessElementMethod(
+        Method targetMethod,
+        int boundedWindowParamIndex,
+        int inputProviderParamIndex,
+        int outputReceiverParamIndex) {
+      this.targetMethod = targetMethod;
+      this.boundedWindowParamIndex = boundedWindowParamIndex;
+      this.inputProviderParamIndex = inputProviderParamIndex;
+      this.outputReceiverParamIndex = outputReceiverParamIndex;
+    }
+
+    /**
+     * Utility method: returns how many parameters the method has, not counting the {@link
+     * DoFn.ProcessContext} parameter.
+     */
+    public int numExtraParameters() {
```

Comment:
I'm not a fan of providing this, especially with this implementation. It seems like it will be high maintenance.